{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4ad3a90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing expression for Organoid,  Thu Mar 31 12:54:27 2022\n",
      "importing pathways,  Thu Mar 31 12:55:10 2022\n",
      "running ssGSEA for reactome ... ,  Thu Mar 31 12:55:12 2022\n"
     ]
    }
   ],
   "source": [
    "############### run single-sample GSEA (ssGSEA) for cancer patient or organoid data ###############\n",
    "import gseapy\n",
    "import gseapy as gp\n",
    "import scipy.stats as stat\n",
    "import numpy as np\n",
    "import time, os\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "with open(\"../utilities/pathway_utilities.py\") as f:\n",
    "    code = compile(f.read(), \"../utilities/pathway_utilities.py\", 'exec')\n",
    "    exec(code, globals())\n",
    "    \n",
    "with open(\"../utilities/Preclinical_model_data.py\") as b:\n",
    "    code = compile(b.read(), \"../utilities/Preclinical_model_data.py\", 'exec')\n",
    "    exec(code, globals())\n",
    "         \n",
    "with open(\"../utilities/patient_expression.py\") as c:\n",
    "    code = compile(c.read(), \"../utilities/patient_expression.py\", 'exec')\n",
    "    exec(code, globals())\n",
    "     \n",
    "g2en=geneID2ensemble()\n",
    "## INITIALIZE\n",
    "#======================\n",
    "# INITIALIZE PARAMETERS\n",
    "source = 'Organoid' # 'organoid', 'TCGA'\n",
    "cancer_type = 'LUAD'\n",
    "testing_pathway_list = ['REACTOME']\n",
    "\n",
    "#==================\n",
    "# IMPORT EXPRESSION\n",
    "print ('importing expression for %s, ' %source, time.ctime())\n",
    "\n",
    "expDic = {} # { sample ID : { gene in uniprot : exp } }\n",
    "expDic_geneID = {} # { sample ID : { gene : exp } }\n",
    "geneList, sampleList = [], []\n",
    "\n",
    "if source.lower() == 'organoid':\n",
    "    expDic_geneID = parse_organoid_transcriptome(cancer_type)\n",
    "#print(expDic_geneID)\n",
    "\n",
    "if source.upper() == 'TCGA':\n",
    "    expDic_geneID = parse_TCGA_log2_FPKM(cancer_type)\n",
    "\n",
    "sampleList = expDic_geneID.keys()\n",
    "sorted(sampleList)\n",
    "\n",
    "for sample in expDic_geneID:\n",
    "    geneList = expDic_geneID[sample].keys()\n",
    "\n",
    "#========================================\n",
    "# IMPORT PATHWAYS FOR ENRICHMENT ANALYSIS\n",
    "print('importing pathways, ', time.ctime())\n",
    "reactome = reactome_genes_ensemble() # { pathway : [ gene list ] }\n",
    "pathwayDic = {'reactome':reactome} # \n",
    "\n",
    "#print(reactome)\n",
    "\n",
    "\n",
    "## PRINT ssGSEA RESULTS\n",
    "\n",
    "#===============\n",
    "# MAKE DIRECTORY\n",
    "fo_directory = 'Result_GSEA' \n",
    "dir_list = [cancer_type.upper(), source]\n",
    "for d in dir_list:\n",
    "    if os.path.isdir('%s/%s' %(fo_directory, d)) == False:\n",
    "        os.mkdir('%s/%s' %(fo_directory, d))\n",
    "        fo_directory = '%s/%s'%(fo_directory, d)\n",
    "\n",
    "#=====================\n",
    "# MAKE GSEA INPUT FILE\n",
    "fiList = os.listdir(fo_directory)\n",
    "\n",
    "# gene expression\n",
    "if not 'expression.txt' in fiList:\n",
    "    fo = open('%s/expression.txt' %(fo_directory), 'w')\n",
    "    d ='\\t'.join(['NAME', 'DESCRIPTION']) + '\\t' + '\\t'.join(sampleList)\n",
    "    print(d, file=fo)\n",
    "    for gene in geneList:\n",
    "        tmp = [gene, 'na']\n",
    "        for sample in sampleList:\n",
    "            tmp.append(expDic_geneID[sample][gene])\n",
    "        a='\\t'.join(map(str, tmp))\n",
    "        print(a, file=fo)\n",
    "    fo.close()\n",
    "\n",
    "\n",
    "#=======\n",
    "# ssGSEA\n",
    "for testing_pathway in testing_pathway_list:\n",
    "    if testing_pathway.lower() in pathwayDic:\n",
    "        print ('running ssGSEA for %s ... , ' %testing_pathway.lower(), time.ctime())\n",
    "        # gene sets for ssGSEA\n",
    "        gene_sets = {}\n",
    "        pw_list = []\n",
    "        for pw in pathwayDic[testing_pathway.lower()]:\n",
    "            for ensemble in pathwayDic[testing_pathway.lower()][pw]:\n",
    "                if not pw in gene_sets:\n",
    "                    gene_sets[pw] = []\n",
    "                gene_sets[pw].append(ensemble)\n",
    "        pw_list = gene_sets.keys()\n",
    "\n",
    "        fo = open('%s/%s.gmt' %(fo_directory, testing_pathway.lower()), 'w')\n",
    "        for pw in pw_list:\n",
    "            m = pw + '\\t' + '\\t'.join(gene_sets[pw])\n",
    "            print(m, file=fo)\n",
    "        fo.close()\n",
    "        gene_sets=reactome_genes_ensemble()\n",
    "\n",
    "        # ssGSEA\n",
    "        ss = gp.ssgsea(data='%s/expression.txt'%(fo_directory), gene_sets='%s/%s.gmt' %(fo_directory, testing_pathway.lower()),outdir='%s/%s_ssgsea_result'%(fo_directory, testing_pathway.lower()),min_size=0,max_size=70000,sample_norm_method='rank',permutation_num=0,no_plot=True,scale=True,)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e86c1dec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed03509",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90fdd987",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
